PPC64: better vthread support for object monitors

reinrich opened the pull request:
Testing
* hotspot:tier1
* test/jdk/jdk/internal/vm/Continuation
* test/jdk/java/lang/Thread/virtual

reinrich committed:
PPC64: better vthread support for object monitors

reinrich committed:
Fix StopThreadTest.java test

TheRealMDoerr commented on src/hotspot/cpu/ppc/macroAssembler_ppc.cpp:
```diff
@@ -2737,6 +2770,7 @@ void MacroAssembler::compiler_fast_unlock_object(ConditionRegister flag, Registe
 void MacroAssembler::compiler_fast_lock_lightweight_object(ConditionRegister flag, Register obj, Register box,
                                                            Register tmp1, Register tmp2, Register tmp3) {
   assert_different_registers(obj, box, tmp1, tmp2, tmp3);
+  assert (UseObjectMonitorTable || tmp3 == noreg, "tmp3 not needed");
```
Whitespace before '('.

TheRealMDoerr commented on src/hotspot/cpu/ppc/macroAssembler_ppc.cpp:
```diff
@@ -4450,8 +4485,18 @@ void MacroAssembler::asm_assert_mems_zero(bool check_equal, int size, int mem_of
   asm_assert(check_equal, msg);
+}
 #endif // ASSERT
+
```
Remark: `asm_assert_not_zero` seems to be unused.

TheRealMDoerr commented on src/hotspot/cpu/ppc/macroAssembler_ppc.cpp:
```diff
@@ -4573,6 +4618,37 @@ void MacroAssembler::zap_from_to(Register low, int before, Register high, int af
   BLOCK_COMMENT("} zap memory region");
 }
 
```
Remark: There's a similar function `clobber_volatile_gprs`.

TheRealMDoerr commented on src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:
```diff
@@ -2202,7 +2212,7 @@ nmethod *SharedRuntime::generate_native_wrapper(MacroAssembler *masm,
   Register r_temp_5     = R26;
   Register r_temp_6     = R27;
-  Register r_return_pc  = R28;
+  Register r_last_java_pc  = R28;
```
Extra whitespace.

TheRealMDoerr commented on src/hotspot/cpu/ppc/register_ppc.hpp:
```diff
@@ -100,6 +100,7 @@ class Register {
   constexpr bool is_valid()       const { return ( 0 <= _encoding && _encoding <  number_of_registers); }
   constexpr bool is_volatile()    const { return ( 0 <= _encoding && _encoding <= 13 ); }
   constexpr bool is_nonvolatile() const { return (14 <= _encoding && _encoding <= 31 ); }
+  constexpr bool is_nonvolatile_accross_preemption() const { return (_encoding == 22) || (_encoding == 31 ); }
```
I'd appreciate a short comment before your new function. The other functions in this section are defined by the ABI. This one is implementation specific to keep values in `TemplateInterpreterGenerator::generate_native_entry` alive after preemption and resuming.

TheRealMDoerr reviewed (commented):
Looks great! I have some minor comments and suggestions. Please note that the fibers branch has new conflicts also due to https://github.com/openjdk/jdk/commit/180affc5718c9bf2f009d6a7aa129cc36335384a (not only). We should probably give a Big Endian platform a quick spin.

reinrich committed:
Fix LM_MONITOR case in compiler_fast_lock_object

reinrich committed:
Feedback Martin

reinrich commented:
Hi Martin,

thanks for the feedback. I've addressed all of it. Hope you're ok with what I've done.
Additionally I fixed a issue with LM_MONITOR. I had a déjà-vu doing it. Maybe it got lost when rebasing on the fibers branch.

> Please note that the fibers branch has new conflicts

I'll take care of that.

Cheers, Richard.

reinrich commented:
Hi Martin,
I'm closing this PR because I want to rebase on the fibers branch. I'll squash all commits for a last downstream review before opening the PR for the upstream loom repo.
Thanks, Richard.
